llvm/lib/Target/EVM/EVMAsmPrinter.cpp

```diff
@@ -323,8 +342,59 @@ void EVMAsmPrinter::emitWideRelocatableSymbol(const MachineInstr *MI) {
   OutStreamer->switchSection(CurrentSection);
 }
 
-void EVMAsmPrinter::emitEndOfAsmFile(Module &) {
+void EVMAsmPrinter::createDataSectionBuffer(const Module &M) {
+  SmallVector<StringRef, 16> UniqueInitializers;
+  SmallVector<std::pair<const GlobalVariable *, StringRef>, 16> Globals;
+  for (const GlobalVariable &GV : M.globals()) {
+    if (GV.getAddressSpace() != EVMAS::AS_CODE || !GV.hasInitializer())
+      continue;
+
+    const auto *CV = dyn_cast<ConstantDataSequential>(GV.getInitializer());
+    if (!CV)
+      continue;
+
+    StringRef Init = CV->getRawDataValues();
+    UniqueInitializers.push_back(Init);
+    Globals.emplace_back(&GV, Init);
+  }
+
+  // Deduplicate initializers.
+  sort(UniqueInitializers);
+  UniqueInitializers.erase(
+      std::unique(UniqueInitializers.begin(), UniqueInitializers.end()),
+      UniqueInitializers.end());
+
+  // Remove initializers that are substrings of others.
+  auto IsSubstring = [&UniqueInitializers](const StringRef &A) {
+    return any_of(UniqueInitializers,
+                  [&A](const StringRef &B) { return A != B && B.contains(A); });
+  };
 
+  UniqueInitializers.erase(remove_if(UniqueInitializers, IsSubstring),
+                           UniqueInitializers.end());
+
+  // Build the data section by concatenating unique initializers.
+  // NOTE: Rather than simply concatenating unique strings, we could attempt
```

Comment:
Certainly an overkill and not a real-life use case.
We can probably try it if it's requested by users one day, maybe some cryptographic libraries can have something alike.